File: crypto.cpp
#include <iostream>
#include <random>
#include <limits>
#include <fstream>
#include <string>
#include "math.h"
#include "crypto.h"
using namespace std;
int generatePrime(int min, int max) {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis(min, max);
    
    int prime;
    do {
        prime = dis(gen);
    } while (!isPrime(prime));
    
    return prime;
}

int generateSecretKey(int p) {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis(2, p - 2);  
    int key;
    do {
        key = dis(gen);
    } while (gcd(key, p - 1) != 1); 
    return key;
}


void shamirProtocol(string& inNameFile) {
    string text;
    ifstream infile(inNameFile, ios::binary);
    if (!infile) {
        cerr << "Ошибка открытия файла!" << endl;
        return;
    }
    char symbol;
    while (infile.get(symbol)) {
        text += symbol;
    }
    infile.close();
    int p = generatePrime(1000, 10000);
    cout << "Простое число p = " << p << endl;
    int Ca = generateSecretKey(p);
    int Da = searchReciprocalNum(Ca, p - 1);
    cout << "Ключи Алисы: Ca = " << Ca << ", Da = " << Da << endl;
    int Cb = generateSecretKey(p);
    int Db = searchReciprocalNum(Cb, p - 1);
    cout << "Ключи Боба: Cb = " << Cb << ", Db = " << Db << endl;
    vector<int> x1Value, x2Values, x3Values,x4Values;
    cout << "\nШаг 1: Алиса вычисляет x1 = m^Ca mod p" << endl;
    for (char c : text) {
        int m = static_cast<int>(c);
        int x1 = powMod(m, Ca, p);
        x1Value.push_back(x1);
    }
    string x1NameFile;
    cout << "Введите имя файла для сохранения x1: ";
    cin >> x1NameFile;
    ofstream outputX1(x1NameFile, ios::binary);
    if (!outputX1) {
        cerr << "Ошибка создания файла x1" << endl;
        return;
    }
    for (int x1 : x1Value) {
        outputX1.write(reinterpret_cast<const char*>(&x1), sizeof(x1));
    }
    outputX1.close();
    cout << "Промежуточный результат x1 сохранен в файл: " << x1NameFile << endl;
    cout << "Шаг 2: Боб вычисляет x2 = x1^Cb mod p" << endl;
    for (int x1 : x1Value) {
        int x2 = powMod(x1, Cb, p);
        x2Values.push_back(x2);
    }
    string x2NameFile;
    cout << "Введите имя файла для сохранения x2: ";
    cin >> x2NameFile;
    ofstream outputX2(x2NameFile, ios::binary);
    if (!outputX2) {
        cerr << "Ошибка создания файла x2!" << endl;
        return;
    }
    for (int x2 : x2Values) {
        outputX2.write(reinterpret_cast<const char*>(&x2), sizeof(x2));
    }
    outputX2.close();
    cout << "Промежуточный результат x2 сохранен в файл: " << x2NameFile << endl;
    cout << "Шаг 3: Алиса вычисляет x3 = x2^Da mod p" << endl;
    for (int x2 : x2Values) {
        int x3 = powMod(x2, Da, p);
        x3Values.push_back(x3);
    }
    string x3NameFile;
    cout << "Введите имя файла для сохранения x3: ";
    cin >> x3NameFile;
    
    ofstream outputX3(x3NameFile, ios::binary);
    if (!outputX3) {
        cerr << "Ошибка создания файла x3!" << endl;
        return;
    }
    for (int x3 : x3Values) {
        outputX3.write(reinterpret_cast<const char*>(&x3), sizeof(x3));
    }
    outputX3.close();
    cout << "Промежуточный результат x3 сохранен в файл: " << x3NameFile << endl;
    cout << "Шаг 4: Боб вычисляет x4 = x3^Db mod p" << endl;
    string decryptedText;
    for (int x3 : x3Values) {
        int x4 = powMod(x3, Db, p);
        x4Values.push_back(x4);
        decryptedText += static_cast<char>(x4);
    }
    string x4NameFile;
    cout << "Введите имя файла для сохранения x4: ";
    cin >> x4NameFile;
    
    ofstream outputX4(x4NameFile, ios::binary);
    if (!outputX4) {
        cerr << "Ошибка создания файла x4!" << endl;
        return;
    }
    for (int x4 : x4Values) {
        outputX4.write(reinterpret_cast<const char*>(&x4), sizeof(x4));
    }
    outputX4.close();
    cout << "Итоговый результат x4 сохранен в файл: " << x4NameFile << endl;
    string decryptedFile;
    cout << "Введите имя файла для сохранения расшифрованного текста: ";
    cin >> decryptedFile;
    
    ofstream decOutput(decryptedFile, ios::binary);
    if (!decOutput) {
        cerr << "Ошибка создания файла с расшифрованным текстом!" << endl;
        return;
    }
    decOutput.write(decryptedText.c_str(), decryptedText.size());
    decOutput.close();
    cout << "Расшифрованный текст сохранен в файл: " << decryptedFile << endl;
}
void mitmAttack() {
    int p = generatePrime(1000, 10000);
    cout << "Простое число p = " << p << endl;
    int Ca = generateSecretKey(p);
    int Da = searchReciprocalNum(Ca, p - 1);
    cout << "Ключи Алисы: Ca = " << Ca << ", Da = " << Da << endl;
    int Cb = generateSecretKey(p);
    int Db = searchReciprocalNum(Cb, p - 1);
    cout << "Ключи Боба: Cb = " << Cb << ", Db = " << Db << endl;
    int Cm = generateSecretKey(p);
    int Dm = searchReciprocalNum(Cm, p - 1);
    cout << "Ключи Мэллари: Cm = " << Cm << ", Dm = " << Dm << endl;
    
    string message;
    cout << "\nВведите сообщение для шифрования: ";
    cin.ignore(numeric_limits<streamsize>::max(), '\n');
    getline(cin, message);
    
    cout << "\n=== ЭМУЛЯЦИЯ АТАКИ MitM ===" << endl;
    cout << "1. Мэллари перехватывает сообщение от Алисы к Бобу" << endl;
    cout << "2. Мэллари подменяет открытые ключи своими" << endl;
    vector<int> intercepted;
    vector<int> malloryDecrypted;
    cout << "\nПроцесс перехвата и подмены:" << endl;
    for (char c : message) {
        int m = static_cast<int>(c);
        int x1 = powMod(m, Ca, p);
        cout << "Алиса -> Мэллари: " << m << "^" << Ca << " mod " << p << " = " << x1 << endl;
        int x2 = powMod(x1, Dm, p); 
        int malloryMessage = powMod(x2, Cm, p); 
        cout << "Мэллари расшифровал: " << x1 << "^" << Dm << " mod " << p << " = " << x2 << endl;
        cout << "Мэллари знает исходное сообщение: " << static_cast<char>(x2) << endl;
        malloryDecrypted.push_back(x2);
        int x3 = powMod(malloryMessage, Cb, p);
        intercepted.push_back(x3);
        cout << "Мэллари -> Бобу: " << malloryMessage << "^" << Cb << " mod " << p << " = " << x3 << endl;
    }
    cout << "\nМэллари перехватил и расшифровал сообщение: ";
    for (int num : malloryDecrypted) {
        cout << static_cast<char>(num);
    }
    cout << endl;
    string bobDecrypted;
    cout << "\nБоб расшифровывает (думая, что это от Алисы):" << endl;
    for (int x : intercepted) {
        int m = powMod(x, Db, p);
        bobDecrypted += static_cast<char>(m);
        cout << "Боб: " << x << "^" << Db << " mod " << p << " = " << m << " -> '" << static_cast<char>(m) << "'" << endl;
    }
    cout << "\nБоб получил (искаженное) сообщение: " << bobDecrypted << endl;
    cout << "Мэллари успешно перехватил и прочитал сообщение!" << endl;
    string modifiedMessage = "hack";
    cout << "Мэллари изменяет сообщение на: " << modifiedMessage << endl;
    vector<int> modifiedEncrypted;
    for (char c : modifiedMessage) {
        int m = static_cast<int>(c);
        int encrypted = powMod(m, Cm, p); 
        encrypted = powMod(encrypted, Cb, p); 
        modifiedEncrypted.push_back(encrypted);
    }
    
    string finalDecrypted;
    for (int x : modifiedEncrypted) {
        int m = powMod(x, Db, p);
        finalDecrypted += static_cast<char>(m);
    }
    
    cout << "Боб получает подмененное сообщение: " << finalDecrypted << endl;
    cout << "✓ Атака MITM завершена успешно!" << endl;
}

File: go/crypto.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"
)

func generatePrime(min, max int) int {
	seed := time.Now().UnixNano()
	r := rand.New(rand.NewSource(seed))
	var prime int
	for {
		prime = r.Intn(max-min+1) + min
		if isPrime(prime) {
			break
		}
	}
	return prime
}

func generateSecretKey(p int) int {
	seed := time.Now().UnixNano() ^ int64(p)
	r := rand.New(rand.NewSource(seed))
	var key int
	for {
		key = r.Intn((p-2)-2+1) + 2
		if gcd(key, p-1) == 1 {
			break
		}
	}
	return key
}

func shamirProtocol(inNameFile string) {
	infile, err := os.Open(inNameFile)
	if err != nil {
		fmt.Println("Ошибка открытия файла!")
		return
	}
	defer infile.Close()

	var sb strings.Builder
	reader := bufio.NewReader(infile)
	for {
		b, err := reader.ReadByte()
		if err != nil {
			break
		}
		sb.WriteByte(b)
	}
	text := sb.String()

	p := generatePrime(1000, 10000)
	fmt.Println("Простое число p =", p)

	Ca := generateSecretKey(p)
	Da := searchReciprocalNum(Ca, p-1)
	fmt.Println("Ключи Алисы: Ca =", Ca, ", Da =", Da)

	Cb := generateSecretKey(p)
	Db := searchReciprocalNum(Cb, p-1)
	fmt.Println("Ключи Боба: Cb =", Cb, ", Db =", Db)

	var x1Values, x2Values, x3Values, x4Values []int

	fmt.Println("\nШаг 1: Алиса вычисляет x1 = m^Ca mod p")
	for _, c := range text {
		m := int(c)
		x1 := powMod(m, Ca, p)
		x1Values = append(x1Values, x1)
	}
	fmt.Print("Введите имя файла для сохранения x1: ")
	var x1NameFile string
	fmt.Scanln(&x1NameFile)
	saveToFileInt(x1NameFile, x1Values)
	fmt.Println("Промежуточный результат x1 сохранен в файл:", x1NameFile)

	fmt.Println("Шаг 2: Боб вычисляет x2 = x1^Cb mod p")
	for _, x1 := range x1Values {
		x2 := powMod(x1, Cb, p)
		x2Values = append(x2Values, x2)
	}
	fmt.Print("Введите имя файла для сохранения x2: ")
	var x2NameFile string
	fmt.Scanln(&x2NameFile)
	saveToFileInt(x2NameFile, x2Values)
	fmt.Println("Промежуточный результат x2 сохранен в файл:", x2NameFile)

	fmt.Println("Шаг 3: Алиса вычисляет x3 = x2^Da mod p")
	for _, x2 := range x2Values {
		x3 := powMod(x2, Da, p)
		x3Values = append(x3Values, x3)
	}
	fmt.Print("Введите имя файла для сохранения x3: ")
	var x3NameFile string
	fmt.Scanln(&x3NameFile)
	saveToFileInt(x3NameFile, x3Values)
	fmt.Println("Промежуточный результат x3 сохранен в файл:", x3NameFile)

	fmt.Println("Шаг 4: Боб вычисляет x4 = x3^Db mod p")
	var decryptedText strings.Builder
	for _, x3 := range x3Values {
		x4 := powMod(x3, Db, p)
		x4Values = append(x4Values, x4)
		decryptedText.WriteRune(rune(x4))
	}
	fmt.Print("Введите имя файла для сохранения x4: ")
	var x4NameFile string
	fmt.Scanln(&x4NameFile)
	saveToFileInt(x4NameFile, x4Values)
	fmt.Println("Итоговый результат x4 сохранен в файл:", x4NameFile)

	fmt.Print("Введите имя файла для сохранения расшифрованного текста: ")
	var decryptedFile string
	fmt.Scanln(&decryptedFile)
	err = os.WriteFile(decryptedFile, []byte(decryptedText.String()), 0644)
	if err != nil {
		fmt.Println("Ошибка создания файла с расшифрованным текстом!")
		return
	}
	fmt.Println("Расшифрованный текст сохранен в файл:", decryptedFile)
}

func mitmAttack() {
	p := generatePrime(1000, 10000)
	fmt.Println("Простое число p =", p)
	Ca := generateSecretKey(p)
	Da := searchReciprocalNum(Ca, p-1)
	fmt.Println("Ключи Алисы: Ca =", Ca, ", Da =", Da)
	Cb := generateSecretKey(p)
	Db := searchReciprocalNum(Cb, p-1)
	fmt.Println("Ключи Боба: Cb =", Cb, ", Db =", Db)
	Cm := generateSecretKey(p)
	Dm := searchReciprocalNum(Cm, p-1)
	fmt.Println("Ключи Екатерины: Cm =", Cm, ", Dm =", Dm)

	fmt.Print("\nВведите сообщение для шифрования: ")
	reader := bufio.NewReader(os.Stdin)
	msgLine, _ := reader.ReadString('\n')
	message := strings.TrimSpace(msgLine)

	fmt.Println("\n=== ЭМУЛЯЦИЯ АТАКИ MitM ===")
	fmt.Println("1. Екатерина перехватывает сообщение от Алисы к Бобу")
	fmt.Println("2. Екатерина подменяет открытые ключи своими")

	var intercepted []int32
	var malloryDecrypted []int32
	fmt.Println("\nПроцесс перехвата и подмены:")
	for _, c := range message {
		m := int(c)
		x1 := powMod(m, Ca, p)
		fmt.Printf("Алиса -> Екатерине: %d ^ %d mod %d = %d\n", m, Ca, p, x1)
		x2 := powMod(x1, Dm, p)
		malloryMessage := powMod(x2, Cm, p)
		fmt.Printf("Екатерина расшифровала: %d ^ %d mod %d = %d\n", x1, Dm, p, x2)
		fmt.Printf("Екатерина знает исходное сообщение: %c\n", byte(x2))
		malloryDecrypted = append(malloryDecrypted, int32(x2))
		x3 := powMod(malloryMessage, Cb, p)
		intercepted = append(intercepted, int32(x3))
		fmt.Printf("Екатерина -> Бобу: %d ^ %d mod %d = %d\n", malloryMessage, Cb, p, x3)
	}

	fmt.Print("\nЕкатерина перехватила и расшифровала сообщение: ")
	for _, num := range malloryDecrypted {
		fmt.Printf("%c", num)
	}
	fmt.Println()

	var bobDecrypted strings.Builder
	fmt.Println("\nБоб расшифровывает (думая, что это от Алисы):")
	for _, x := range intercepted {
		m := powMod(int(x), Db, p)
		bobDecrypted.WriteByte(byte(m))
		fmt.Printf("Боб: %d ^ %d mod %d = %d -> '%c'\n", x, Db, p, m, byte(m))
	}

	fmt.Println("\nБоб получил (искаженное) сообщение:", bobDecrypted.String())
	fmt.Println("Екатерина успешно перехватила и прочитала сообщение!")
	modifiedMessage := "hack"
	fmt.Println("Екатерина изменяет сообщение на:", modifiedMessage)

	var modifiedEncrypted []int32
	for _, c := range modifiedMessage {
		m := int(c)
		encrypted := powMod(m, Cm, p)
		encrypted = powMod(encrypted, Cb, p)
		modifiedEncrypted = append(modifiedEncrypted, int32(encrypted))
	}

	var finalDecrypted strings.Builder
	for _, x := range modifiedEncrypted {
		m := powMod(int(x), Db, p)
		finalDecrypted.WriteByte(byte(m))
	}

	fmt.Println("Боб получает подмененное сообщение:", finalDecrypted.String())
	fmt.Println("✓ Атака MITM завершена успешно!")
}
func saveToFileInt(filename string, values []int) {
	outfile, err := os.Create(filename)
	if err != nil {
		fmt.Println("Ошибка создания файла:", filename)
		return
	}
	defer outfile.Close()

	for _, v := range values {
		_ = binary.Write(outfile, binary.LittleEndian, int32(v))
	}
}
